File: top_interview_questions/java/78_subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList ();
        for (int i = 0; i <= nums.length; i++)
            backtrack(ans, nums, new ArrayList<Integer> (), 0, i);
        return ans;
    }

    public void backtrack(List<List<Integer>> ans, int[] nums, ArrayList<Integer> temp, int start, int k) {
        if (temp.size() == k)
            ans.add(new ArrayList(temp));
        for (int i = start; i < nums.length; i++) {
            temp.add(nums[i]);
            backtrack(ans, nums, temp, i+1, k);
            temp.remove(temp.size()-1);
        }
    }
}

File: top_interview_questions/python/78_subsets.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: Fizz Cao 
@file: 78_subsets.py
@time: 22/09/2021 10:58 pm
'''
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        for i in range(len(nums) + 1):
            self.backtrack(ans, nums, [], i, 0)
        return ans

    def backtrack(self, ans, nums, temp, k, start):
        if len(temp) == k:
            ans.append(temp)
        for i in range(start, len(nums)):
            self.backtrack(ans, nums, temp + [nums[i]], k, i + 1)